Blog: Using TimescaleDB with Node-RED (#2116)
FlowFuse/website: src/blog/2024/06/using-timescaledb-with-node-red.md

---
title: How to use TimescaleDB with Node-RED
subtitle: Guide to integrating TimescaleDB into Node-RED for Storing, Updating, and Querying Data.
description: Learn how to integrate TimescaleDB with Node-RED for storing and managing time-series data efficiently.
date: 2024-06-19
authors: ["sumit-shinde"]
image: /blog/2024/06/images/timescaledb-with-node-red.png
tags:
   - posts
   - node-red
   - timescaledb
   - postgresql
   - time-series database
---

In the context of IoT and IIoT applications, time series databases are essential for storing data based on timestamps. While InfluxDB has been a popular choice for a long time, another time series database, TimescaleDB, is gaining popularity. This article will cover how to use TimescaleDB with Node-RED, how TimescaleDB works, and the queries needed when building IoT applications.

> If you prefer video tutorials, a few months ago, Gray, OT Data & Community Strategist at Flowfuse, conducted a [live session on TimescaleDB](https://www.youtube.com/watch?v=MD1U6LDqJ1c).

Reviewer: ```suggestion
If you prefer video tutorials, a few months ago, Grey, OT Data & Community Strategist at Flowfuse, conducted a [live session on TimescaleDB](https://www.youtube.com/watch?v=MD1U6LDqJ1c).
```

2. Once logged in, create a new TimescaleDB service by following the on-screen instructions.
3. Choose your service settings, such as region, CPU, memory, and storage requirements, based on your application's needs.
4. After creating the service, you’ll see the connection details. If you cannot see them, go to the "Services" option in the sidebar, click on the created service, and then in the "Overview" tab at the bottom, you will see your configuration details.

## Using TimescaleDB with Node-RED

In this section of the guide, we will explore integrating TimescaleDB with Node-RED. We'll cover creating and deleting Hypertables, and inserting, updating, and deleting data from these tables. Additionally, we'll delve into advanced queries for comprehensive data analysis. Throughout this guide, we'll use a temperature example to illustrate each operation.

### Installing PostgreSQL Custom Node

Since TimescaleDB is built on top of PostgreSQL, we can use the PostgreSQL node.

1. Click the Node-RED Settings (top-right).
2. Click "Manage Palette".
3. Search for `node-red-contrib-postgresql`.
4. Click "Install".

### Configuring PostgreSQL node with TimescaleDB configurations

> Before proceeding further make sure you have added environment variables for your TimescaleDB configuration details, for more information refer to [Using environment variables with Node-RED](/blog/2023/01/environment-variables-in-node-red/).

Reviewer: ```suggestion
Before proceeding, make sure you have added environment variables for your TimescaleDB configuration details. For more information, refer to [Using environment variables with Node-RED](/blog/2023/01/environment-variables-in-node-red/).
```

date: 2024-06-03
authors: ["sumit-shinde"]
image: 
tags:
   - posts
   - node-red
   - timescaledb
   - postgresql
   - time-series database
---

In the context of IoT and IIoT applications, time series databases are essential for storing data based on timestamps. While InfluxDB has been a popular choice for a long time, another time series database, TimescaleDB, is gaining popularity. This article will cover how to use TimescaleDB with Node-RED, when to choose it over InfluxDB, and important operations and queries needed when building IoT applications.

If you prefer video tutorials, a few months ago, Gray, OT Data & Community Strategist at Flowfuse, conducted a live session on TimescaleDB. You can watch it [here](https://www.youtube.com/watch?v=MD1U6LDqJ1c).

<!--more-->

## What is TimeScaleDB

TimescaleDB is a time-series database built on PostgreSQL for handling large time-stamped data efficiently. It's optimized for time-series data, which offers automatic partitioning, optimized indexing, compression, and advanced analytics like continuous aggregates. Unlike PostgreSQL, it focuses specifically on time-series data management with distributed hypertable architecture.

> ## TimescaleDB Vs InfluxDB

Reviewer: While I like the comparison between the two, I am curious if we want to get into the game of saying which is better or worse. Even though that isn't what you are saying, it will be interpreted that way by some.

I will defer to @ZJvandeWeg for guidance.

Author: Should we remove the comparison section?

> TimescaleDB is a time-series database built on PostgreSQL for handling large time-stamped data efficiently. It's optimized for time-series data, which offers automatic partitioning, optimized indexing, compression, and advanced analytics like continuous aggregates. Unlike PostgreSQL, it focuses specifically on time-series data management with distributed hypertable architecture.

Author: Yes, I have explained each of them in their own sections below in  but I will move that here or update it to explain how TimescaleDB works.

> TimescaleDB is a time-series database built on PostgreSQL for handling large time-stamped data efficiently. It's optimized for time-series data, which offers automatic partitioning, optimized indexing, compression, and advanced analytics like continuous aggregates. Unlike PostgreSQL, it focuses specifically on time-series data management with distributed hypertable architecture.

Author: I have updated the section and requested this visual https://github.com/FlowFuse/website/issues/2202

| **Scalability & HA** | Multi-node clustering, read replicas, automated high availability with multi-AZ deployments.                  | Clustering and high availability are offered, but TimescaleDB provides more robust solutions for enterprise use.   |
| **Geospatial Support** | Supports geospatial and vector data types, ideal for geographical analysis and machine learning.              | Focused primarily on time-series data, lacking native support for advanced data types.                         |
| **Disk Usage** | Effective compression with lower memory usage, narrowing storage gap at higher cardinalities.                 | Achieves better compression ratios with column-oriented structure, especially at lower cardinalities.           |
| **CPU Usage** | 10x better resource utilization according to DNSFilter comparison, even with 30% higher requests.            | Higher CPU usage compared to TimescaleDB.                                                                     |

## Setting up TimescaleDB environment

### Installing TimescaleDB locally

> If you want to install TimescaleDB locally on your system, you can follow their official documentation on [Install TimescaleDB](https://docs.timescale.com/self-hosted/latest/install/).

Reviewer: ```suggestion
If you want to install TimescaleDB locally, you can follow their official documentation on [Install TimescaleDB](https://docs.timescale.com/self-hosted/latest/install/).
```

### Using TimescaleDB cloud option

TimescaleDB also offers a cloud option that simplifies deployment and management. Here’s how to set it up:

1. Go to the [Timescale Cloud](https://console.cloud.timescale.com/signup) website and sign up for an account.
2. Once logged in, create a new TimescaleDB service by following the on-screen instructions.
3. Choose your service settings, such as region, CPU, memory, and storage requirements, based on your application's needs.

> 4. After the service is created, you’ll see the connection details. If you cannot see them, go to the "Services" option in the sidebar, click on the created service, and then in the "Overview" tab at the bottom, you will see your configuration details.

Reviewer: ```suggestion
4. After creating the service, you’ll see the connection details. If you cannot see them, go to the "Services" option in the sidebar, click on the created service, and then in the "Overview" tab at the bottom, you will see your configuration details.
```

{% endrenderFlow %}

### Inserting Data into the Table

> To insert data into a TimescaleDB Hypertable, the steps are similar to inserting data into a standard PostgreSQL table.

Reviewer: ```suggestion
The steps to insert data into a TimescaleDB Hypertable are similar to inserting data into a standard PostgreSQL table.
```

{% endrenderFlow %}

## Deploying the Flow

1. To test the imported flows, you need to deploy them. To do that, click on the deploy button located in the top right corner.

> After deploying the flow, you can test each operation such as creating, deleting, updating, and other queries by clicking on the inject button. Upon successful operation, you will be able to see the results in the debug panel of the sidebar. Additionally, if you want to learn more about PostgreSQL, you can refer to the [Using PostgreSQL with Node-RED](/blog/2024/02/postgresql-with-node-red/) where you will also find the section which shows the messages received after a successful operation by the PostgresWQL node.

Reviewer: ```suggestion
After deploying the flow, you can test each operation such as creating, deleting, updating, and other queries by clicking on the inject button. Upon successful operation, you will be able to see the results in the debug panel of the sidebar. If you want to learn any additional information about PostgreSQL, you can refer to the [Using PostgreSQL with Node-RED](/blog/2024/02/postgresql-with-node-red/) where you will also find the section which shows the messages received after a successful operation by the PostgresWQL node.
```

## Conclusion

> In this guide, we have demonstrated how to integrate TimescaleDB with Node-RED. We have covered various operations, including creating hypertables, deleting them, and performing data manipulation tasks such as inserting, updating, and deleting data and executing advanced queries.

Reviewer: ```suggestion
In this guide, we have demonstrated how to integrate TimescaleDB with Node-RED and covered various operations, including creating hypertables and deleting them, performing data manipulation tasks such as inserting, updating, and deleting data, and executing advanced queries.
```
